controller/scheduler-v2/scheduler.go

```diff
@@ -62,63 +69,92 @@ func (s *Scheduler) Run() error {
 	log.Info("starting scheduler loop")
 	defer log.Info("exiting scheduler loop")
 
-	for {
-		// first, check if we should stop or process pending job events
-		select {
-		case <-s.stop:
-			return nil
-		case req := <-s.jobRequests:
-			s.HandleJobRequest(req)
-			continue
-		default:
+	jobTicker := time.Tick(30 * time.Second)
+	formationTicker := time.Tick(time.Minute)
+	go func() {
+		for {
+			select {
+			case <-jobTicker:
+				s.jobSync <- struct{}{}
+			case <-formationTicker:
+				s.formationSync <- struct{}{}
+			}
 		}
+	}()
+	s.jobSync <- struct{}{}
+	s.formationSync <- struct{}{}
+
+	chanList := []*chanHandler{
+		{
+			channel: s.jobRequests,
+			handler: s.jobRequestHandler,
+		}, {
+			channel: s.rectifyJobs,
+			handler: s.rectifyJobsHandler,
+		}, {
+			channel: s.formationChange,
+			handler: s.formationChangeHandler,
+		}, {
+			channel: s.formationSync,
+			handler: s.formationSyncHandler,
+		}, {
+			channel: s.jobSync,
+			handler: s.jobSyncHandler,
+		},
+	}
 
-		log.Info("starting cluster sync")
-		if err := s.Sync(); err != nil {
-			log.Error("error performing cluster sync", "err", err)
-			continue
+	for {
+		if s.handleChan(&chanHandler{s.stop, nil}) {
+			break
 		}
-
-		log.Info("starting watching events")
-		select {
-		case <-s.stop:
-			return nil
-		case fc := <-s.formationChange:
-			if err := s.FormationChange(fc); err != nil {
-				log.Error("error performing formation change", "err", err)
-				continue
+		handled := false
+		for _, chHandler := range chanList {
+			handled = s.handleChan(chHandler)
```

Comment:
1) Several events can conflict. E.g. jobSync reveals jobs with unknown formations, triggers formation sync. Formation sync pulls down the new formations, which triggers a jobSync and a rectify. If the rectify happens before the jobSync, it will launch a new job instead of syncing to the existing one.
2) Simplifies testing.